phe/util.py

```diff
@@ -170,3 +174,247 @@ def int_to_base64(source):
+    16139, 16141, 16183, 16187, 16189, 16193, 16217, 16223, 16229, 16231,
+    16249, 16253, 16267, 16273, 16301, 16319, 16333, 16339, 16349, 16361,
+    16363, 16369, 16381, 16411, 16417, 16421, 16427, 16433, 16447, 16451,
+    16453, 16477, 16481, 16487, 16493, 16519, 16529, 16547, 16553, 16561,
+    16567, 16573, 16603, 16607, 16619, 16631, 16633, 16649, 16651, 16657,
+    16661, 16673, 16691, 16693, 16699, 16703, 16729, 16741, 16747, 16759,
+    16763, 16787, 16811, 16823, 16829, 16831, 16843, 16871, 16879, 16883,
+    16889, 16901, 16903, 16921, 16927, 16931, 16937, 16943, 16963, 16979,
+    16981, 16987, 16993, 17011, 17021, 17027, 17029, 17033, 17041, 17047,
+    17053, 17077, 17093, 17099, 17107, 17117, 17123, 17137, 17159, 17167,
+    17183, 17189, 17191, 17203, 17207, 17209, 17231, 17239, 17257, 17291,
+    17293, 17299, 17317, 17321, 17327, 17333, 17341, 17351, 17359, 17377,
+    17383, 17387, 17389, 17393, 17401, 17417, 17419, 17431, 17443, 17449,
+    17467, 17471, 17477, 17483, 17489, 17491, 17497, 17509, 17519, 17539,
+    17551, 17569, 17573, 17579, 17581, 17597, 17599, 17609, 17623, 17627,
+    17657, 17659, 17669, 17681, 17683, 17707, 17713, 17729, 17737, 17747,
+    17749, 17761, 17783, 17789, 17791, 17807, 17827, 17837, 17839, 17851,
+    17863,
+]
+
+
+def miller_rabin(n, k):
+    """Run the Miller-Rabin test on n with at most k iterations
+
+    Arguments:
+        n (int): number whose primality is to be tested
+        k (int): maximum number of iterations to run
+
+    Returns:
+        bool: If n is prime, then True is returned. Otherwise, False is
+        returned, except with probability less than 4**-k.
+
+    See <https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test>
+    """
+    assert n > 3
+
+    # find r and d such that n-1 = 2^r × d
+    d = n-1
+    r = 0
+    while d % 2 == 0:
+        d //= 2
+        r += 1
+    assert n-1 == d * 2**r
+    assert d % 2 == 1
+
+    for _ in range(k):  # each iteration divides risk of false prime by 4
+        a = random.randint(2, n-2)  # choose a random witness
+
+        x = pow(a, d, n)
+        if x == 1 or x == n-1:
+            continue  # go to next witness
+
+        for _ in range(1, r):
+            x = x*x % n
+            if x == n-1:
+                break   # go to next witness
+        else:
+            return False
+    return True
+
+
+def is_prime(n):
+    """Test whether n is probably prime
+
+    See <https://en.wikipedia.org/wiki/Primality_test#Probabilistic_tests>
+    """
+    # as an optimization we quickly detect small primes using the list above
+    if n <= first_primes[-1]:
+        return n in first_primes
+    # for small dividors (relatively frequent), euclidean division is best
+    for p in first_primes:
+        if n % p == 0:
+            return False
+    # the actual generic test; give a false prime with probability 2⁻⁵⁰
+    return miller_rabin(n, 25)
```

Comment:
The part which took us longer than expected: this number 25 related to the number of iterations to run for Miller-Rabin.
Because this is used to generate cryptographic keys, it needs to be satisfying some properties.
We found the following document:
https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf#page=80
so because this parameter may vary depending on key sizes or security required, it would be good to be able to parametrize it. 
As this is not a number used in the other libraries (e.g. if GMP is there), we could configure it as an environment variable,having as a default value 25 which seems sensible.

Reply:
Parametrizing the number of Miller-Rabin rounds outside the program is a bit unusual. The NIST recommendations seems a *tad* paranoid, especially once you take into account that the probability of miscomputation. In particular, see the first two answers to this [SO question](https://security.stackexchange.com/questions/4544/how-many-iterations-of-rabin-miller-should-be-used-to-generate-cryptographic-saf). The most practical approach might be to follow Thomas Pornin's advice and use 40 iterations which, albeit large, is not *too* costly, safe, and avoid introducing an environment variable.

However, if you *do* want to make it a parameter anyway, what would be your preferred approach, since the project does not use other environment variables? I am thinking of something akin to `def is_prime(n, nr_rounds=None):`, then, if `nr_round` is `None`, load `os.environ['SOMETHING']` (cache result in global variable). Since that would be the only use of an environment variable in the project, I think it would be better not to introduce new functions or modules just for this feature (if other environment variables are introduced, this may be changed).

Reply:
That also reminds me that you might want to use [safe primes](https://en.wikipedia.org/wiki/Safe_prime) at key generation. That should probably go into a separate pull request though.

Reply:
Thanks for the discussion, and using a defensive default value is always better so let's go for 40.

However, we should really let the user of the library being able to select this number (for expert user). And I totally agree with you that we should not introduce a compulsory environment variable neither new functions nor modules. 
Just for the discussion, we could also have `def is_prime(n, nr_rounds=40):` explaining why the default value would be 40, but then in the method `getprimeover(N):` (which is the only method using it currently), check if `os.environ['SOMETHING']` exists and use it if available.

Reply:
I am still skeptical that adding an environment variable to parametrize the number of rounds is a good idea. Even very generic library in wide use to not offer this feature.

GMP always uses 25 iterations, and [*Handbook of Applied Cryptography*](http://cacr.uwaterloo.ca/hac/about/chap4.pdf) (page 17 of PDF, page 148 of book, figure 4.4) only recommends 2 iterations, which is what OpenSSL follows (see `include/openssl/bn.h`, macro `BN_prime_checks_for_size()` used in `BN_generate_prime_ex()` and `BN_is_prime_fasttest_ex()` of `crypto/bn/bn_prime.c`). SymPy does uses 46 iterations (in `ntheory/primetest.py`, `isprime()` calls `mr()` with the first 46 prime numbers as bases), although I am not sure why they use prime numbers as bases.

My impression is that 2 iterations are already enough but low cost and paranoia means that the number of iterations can be raise to a few dozens. After having checked the behavior of GMP (since `phe` uses `gmpy2` if available), I think it might actually be best to set the number of rounds to 25 for consistency. Choosing 40 rounds rather than 25 would not change much, but I advise against adding an environment variable.

I will follow your preferences regarding the approach to follow in `phe`, but I prefer to first make sure this is what you want to do.

Reply:
Thanks for checking all these libraries. 
And thanks a lot for your advises which make a lot of sense.
As our preferred use of the library is with GMP, following their defaults is the most sensible.
So let's go for 25 iterations, and no environment variable. 
We should just document this choice by pointing out that GMP is always using this value.

Reply:
Thanks for taking my arguments into consideration!